app/src/main/java/net/gsantner/markor/frontend/textview/HighlightingEditor.java

```diff
@@ -546,4 +463,181 @@ public int setSelectionExpandWholeLines() {
+                return true;
+            }
+        }
+
+        /**
+         * Draw line numbers.
+         *
+         * @param canvas the canvas on which the line numbers will be drawn.
+         */
+        public void draw(final Canvas canvas) {
+            if (!_editor.getLocalVisibleRect(_visibleArea)) {
+                return;
+            }
+
+            final CharSequence text = _editor.getText();
+            final Layout layout = _editor.getLayout();
+            if (text == null || layout == null) {
+                return;
+            }
+
+            final int offsetY = _editor.getPaddingTop();
+            final int layoutLineCount = layout.getLineCount();
+
+            // Only if the visible area is out of current line numbers area or the layout line count changed
+            // We make a single pass through the text to determine
+            // 1. Start line number to be drawn at present;
+            // 2. Index of layout lines to be drawn at present.
+            if (isOutOfLineNumbersArea() || layoutLineCount != _lastLayoutLineCount) {
+                _lastLayoutLineCount = layoutLineCount;
+                _startNumber = 1;
+                _lineIndex.clear();
+                _top = _visibleArea.top - _visibleArea.height();
+                _bottom = _visibleArea.bottom + _visibleArea.height();
+                for (int i = 0; i < layoutLineCount; i++) {
+                    final int start = layout.getLineStart(i);
+                    if (start == 0 || text.charAt(start - 1) == '\n') {
+                        final int y = layout.getLineBaseline(i);
+                        if (y > _bottom) {
+                            break;
+                        } else if (y < _top) {
+                            _startNumber++;
+                        } else {
+                            _lineIndex.add(i);
+                        }
+                    }
+                }
+            }
+
+            // Only if the text size or the max line number of digits changed
+            // Update the gutter parameters and set padding
+            if (_paint.getTextSize() != _lastTextSize || getNumberDigits(_maxNumber) != _lastMaxNumberDigits) {
+                final int width = Math.round(_paint.measureText(String.valueOf(_maxNumber)));
+                _numberX = LINE_NUMBER_PADDING_LEFT + width;
+                _gutterLineX = _numberX + LINE_NUMBER_PADDING_RIGHT;
+                _editor.setPadding(_gutterLineX + 10, _editor.getPaddingTop(), _editor.getPaddingRight(), _editor.getPaddingBottom());
+            }
+
+            // Draw the gutter line
+            _paint.setColor(Color.LTGRAY);
+            canvas.drawLine(_gutterLineX, _top, _gutterLineX, _bottom, _paint);
+
+            // Draw the line numbers
+            _paint.setColor(Color.GRAY);
+            final int size = _lineIndex.size();
+            for (int i = 0; i < size; i++) {
```

Comment:
The reason I had 2 loops with an ArrayList is because the first loop computes both the maxLineNumber and the number y positions.

If we are going to compute max line number in a text watcher, then this loop can be combined with the first one and we can eliminate the need for the ArrayList and _startNumber etc etc